packages/access-api/src/service/upload-api-proxy.js

```diff
@@ -0,0 +1,266 @@
+  }
+  return store
+}
+
+/**
+ * @template {Record<string, any>} T
+ * @param {object} options
+ * @param {iucanto.Signer} [options.signer]
+ * @param {Pick<Map<dagUcan.DID, iucanto.ConnectionView<T>>, 'get'>} options.connections
+ */
+function createProxyUploadService(options) {
+  const handleInvocation = createInvocationResponder(options)
+  /**
+   * @type {UploadService}
+   */
+  const store = {
+    add: handleInvocation,
+    list: handleInvocation,
+    remove: handleInvocation,
+  }
+  return store
+}
+
+/**
+ * @typedef UcantoHttpConnectionOptions
+ * @property {dagUcan.DID} audience
+ * @property {typeof globalThis.fetch} options.fetch
+ * @property {URL} options.url
+ */
+
+/**
+ * @implements {Pick<Map<dagUcan.DID, iucanto.Connection<any>>, 'get'>}
+ */
+class AudienceConnections {
+  /** @type {Record<dagUcan.DID, URL>} */
+  #audienceToUrl
+  /** @type {undefined|iucanto.ConnectionView<any>} */
+  #defaultConnection
+  /** @type {typeof globalThis.fetch} */
+  #fetch
+
+  /**
+   * @param {UcantoHttpConnectionOptions} options
+   * @returns {iucanto.ConnectionView<any>}
+   */
+  static createConnection(options) {
+    return Client.connect({
+      id: dagUcanDid.parse(options.audience),
+      encoder: CAR,
+      decoder: CBOR,
+      channel: HTTP.open({
+        fetch: options.fetch,
+        url: options.url,
+      }),
+    })
+  }
+
+  /**
+   * @param {object} options
+   * @param {Record<dagUcan.DID, URL>} options.audienceToUrl
+   * @param {UcantoHttpConnectionOptions} [options.defaultConnection]
+   * @param {typeof globalThis.fetch} options.fetch
+   */
+  constructor(options) {
+    this.#fetch = options.fetch
+    this.#audienceToUrl = options.audienceToUrl
+    this.#defaultConnection = options.defaultConnection
+      ? AudienceConnections.createConnection(options.defaultConnection)
+      : undefined
+  }
+
+  /**
+   * @param {dagUcan.DID} audience
+   */
+  get(audience) {
+    if (audience in this.#audienceToUrl) {
+      return AudienceConnections.createConnection({
+        audience,
+        fetch: this.#fetch,
+        url: this.#audienceToUrl[audience],
+      })
+    }
+    return this.#defaultConnection
+  }
+}
+
+export class UploadApiProxyService {
+  /** @type {StoreService} */
+  store
+  /** @type {UploadService} */
+  upload
+
+  /**
+   * @type {Record<string, {
+   * audience: dagUcan.DID,
+   * url: URL,
+   * }>}
+   */
+  static #environments = {
+    production: {
+      audience: 'did:web:web3.storage',
+      url: new URL('https://up.web3.storage'),
+    },
+    staging: {
+      audience: 'did:web:staging.web3.storage',
+      url: new URL('https://staging.up.web3.storage'),
+    },
+  }
+
+  /**
+   * @param {object} options
+   * @param {iucanto.Signer} [options.signer]
+   * @param {typeof globalThis.fetch} options.fetch
+   */
+  static create(options) {
+    // eslint-disable-next-line unicorn/no-array-reduce
+    const audienceToUrl = Object.values(this.#environments).reduce(
```

Comment:
Do we ever need to have multiple connections in practice ? I'd expect we use one in one staging and the other in production, if so it seems like there's bunch of unnecessary code to be maintained with added complexity costs.

As a side note we should probably add code coverage which would decentivize adding code which isn't used.

Reply:
> Do we ever need to have multiple connections in practice ?

No. Though we could also ask "Do we ever need a unified access-api + upload-api single environment service in practice?". Not yet, but I think we both believe it will make working with the total API more intuitive (i.e. api user won't have to worry about which service to invoke).

When we talked I like your vision of 'If you send an invocation to any API, it should try to route it to a place that can handle it'. i.e. unifying api across services. Doing it this way also unifies across environments. I don't think it's strictly necessary, but I do believe it will make working with the overall platform slightly more intuitive. (i.e. api user won't have to worry about **which environment's** 'unified' api to invoke).

I definitely agree code coverage could be useful. https://github.com/web3-storage/w3protocol/issues/339#issue-1531511259